owltorrent-network/src/main/java/edu/rice/owltorrent/network/SocketConnector.java

```diff
@@ -0,0 +1,136 @@
+              log.error(e);
+            }
+          }
+        }
+      };
+
+  private SocketConnector(
+      Peer peer,
+      NetworkToStorageAdapter storageAdapter,
+      MessageReader messageReader,
+      Socket peerSocket) {
+    super(peer, storageAdapter, messageReader);
+    this.peerSocket = peerSocket;
+  }
+
+  public static SocketConnector initiateConnection(
+      Peer peer, NetworkToStorageAdapter storageAdapter) throws IOException {
+    SocketConnector connector =
+        new SocketConnector(
+            peer,
+            storageAdapter,
+            new BusyWaitMessageReader(),
+            new Socket(peer.getAddress().getAddress(), peer.getAddress().getPort()));
+    connector.initiateConnection();
+    return connector;
+  }
+
+  public static SocketConnector respondToConnection(
+      Peer peer, Socket peerSocket, NetworkToStorageAdapter storageAdapter) throws IOException {
+    SocketConnector connector =
+        new SocketConnector(peer, storageAdapter, new BusyWaitMessageReader(), peerSocket);
+    connector.respondToConnection();
+    return connector;
+  }
+
+  @Override
+  protected void initiateConnection() throws IOException {
+    this.in = new DataInputStream(peerSocket.getInputStream());
+    this.out = new DataOutputStream(peerSocket.getOutputStream());
+
+    this.out.write(constructHandShakeMessage(this.peer));
+
+    // read and confirm handshake from peer
+    byte[] incomingHandshakeBuffer = new byte[68];
+    int readByteLength = in.read(incomingHandshakeBuffer);
+    if (readByteLength != 68 || !confirmHandShake(incomingHandshakeBuffer, this.peer)) {
+      throw new IOException(
+          String.format("Invalid handshake from peer id=%s", this.peer.getPeerID().getId()));
+    }
+    // listen for input with busy waiting
+    new Thread(listenForInput).start();
+  }
+
+  @Override
+  protected void respondToConnection() throws IOException {
+    this.in = new DataInputStream(peerSocket.getInputStream());
+    this.out = new DataOutputStream(peerSocket.getOutputStream());
+
+    this.out.write(constructHandShakeMessage(this.peer));
+    // listen for input with busy waiting
+    new Thread(listenForInput).start();
+  }
+
+  @Override
+  public void writeMessage(PeerMessage message) throws IOException {
+    out.write(message.toBytes());
+  }
+
+  private static byte[] constructHandShakeMessage(Peer peer) {
+    ByteBuffer message = ByteBuffer.allocate(68);
+    message.put((byte) 19);
+    byte[] pstr = new String("BitTorrent protocol").getBytes();
+    message.put(pstr);
+    message.put(new byte[8]);
+    message.put(peer.getTorrent().getInfoHash());
+    message.put(peer.getPeerID().getBytes());
+    return message.array();
+  }
+
+  private static boolean confirmHandShake(byte[] buffer, Peer peer) {
+    if (buffer[0] != 19) return false;
+
+    byte[] title = "BitTorrent protocol".getBytes();
+    for (int i = 1; i < 20; i++) if (title[i - 1] != buffer[i]) return false;
```

Comment:
same here, would be ideal if these were all named constants, + maybe a comment reference to the bittorrent specification and line number